pkg/testutils/cluster.go

```diff
@@ -0,0 +1,202 @@
+// +build functional
+
+//
+// Copyright (c) 2018 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), as published by the Free Software Foundation,
+// or under the Apache License, Version 2.0 <LICENSE-APACHE2 or
+// http://www.apache.org/licenses/LICENSE-2.0>.
+//
+// You may not use this file except in compliance with those terms.
+//
+
+package testutils
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	client "github.com/heketi/heketi/client/api/go-client"
+	"github.com/heketi/heketi/pkg/glusterfs/api"
+	"github.com/heketi/heketi/pkg/utils"
+
+	"github.com/heketi/tests"
+)
+
+type ClusterEnv struct {
+	Nodes   []string
+	SSHPort string
+	Disks   []string
+
+	HeketiUrl string
+}
+
+func (ce *ClusterEnv) Update() {
+	// update the port from the env
+	env := os.Getenv("HEKETI_TEST_STORAGEPORT")
+	if "" != env {
+		ce.SSHPort = env
+	}
+
+	// update the node IPs from the env
+	for i := 0; i < len(ce.Nodes); i++ {
+		key := fmt.Sprintf("HEKETI_TEST_STORAGE%v", i)
+		value := os.Getenv(key)
+		if "" != value {
+			ce.Nodes[i] = value
+		}
+	}
+}
+
+func (ce *ClusterEnv) client() *client.Client {
+	return client.NewClientNoAuth(ce.HeketiUrl)
+}
+
+func (ce *ClusterEnv) SshHost(index int) string {
+	return ce.Nodes[index] + ":" + ce.SSHPort
+}
+
+func (ce *ClusterEnv) Setup(t *testing.T, numNodes, numDisks int) {
+	heketi := ce.client()
+
+	// As a testing invariant, we always expect to set up a cluster
+	// at the start of a test on a _clean_ server.
+	// Verify that there are no outstanding operations on the
+	// server. A test that needs to mess with the operations _must_
+	// clean up after itself.
+	oi, err := heketi.OperationsInfo()
+	tests.Assert(t, err == nil, "expected err == nil, got:", err)
+	tests.Assert(t, oi.Total == 0, "expected oi.Total == 0, got", oi.Total)
+	tests.Assert(t, oi.InFlight == 0, "expected oi.InFlight == 0, got", oi.Total)
+
+	// Create a cluster
+	cluster_req := &api.ClusterCreateRequest{
```

Comment:
to make `Setup` function more useful, how about making `Block` and `File` as input parameters?

Reply:
yeah, I think we can further generalize later..